mrmustard/lab_dev/transformations/transformations.py

```diff
@@ -0,0 +1,320 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The classes representing transformations in quantum circuits.
+"""
+
+from __future__ import annotations
+
+from typing import Optional, Iterable, Tuple, Union
+
+from mrmustard import math
+from .base import Unitary, Channel
+from ...physics.representations import Bargmann
+from ...physics import triples
+from ..utils import make_parameter, reshape_params
+
+__all__ = ["Attenuator", "BSgate", "Dgate", "Sgate"]
+
+
+class BSgate(Unitary):
+    r"""The beam splitter gate.
+
+    It applies to a single pair of modes.
+
+    .. code-block ::
+
+        >>> import numpy as np
+        >>> from mrmustard.lab_dev import BSgate
+
+        >>> unitary = BSgate(modes=[1, 2], theta=0.1)
+        >>> assert unitary.modes == [1, 2]
+        >>> assert np.allclose(unitary.theta.value, 0.1)
+        >>> assert np.allclose(unitary.phi.value, 0.0)
+
+    Args:
+        modes: The modes this gate is applied to.
+        theta: The transmissivity angle.
+        theta_bounds: The bounds for the transmissivity angle.
+        theta_trainable: Whether theta is a trainable variable.
+        phi: The phase angle.
+        phi_bounds: The bounds for the phase angle.
+        phi_trainable: Whether phi is a trainable variable.
+
+    Raises:
+        ValueError: If ``modes`` is not a pair of modes.
+
+    .. details::
+
+        The beamsplitter gate is a Gaussian gate defined by
+
+        .. math::
+            S &= \begin{bmatrix}
+                    \text{Re}(U) & -\text{Im}(U)\\
+                    \text{Im}(U) & \text{Re}(U)
+                \end{bmatrix} \\
+            d &= O_4\:,
+
+        with
+
+        .. math::
+            U &= \begin{bmatrix}
+                    \text{cos}(\theta) & -e^{-i\phi}\text{sin}(\theta)\\
+                    e^{i\phi}\text{sin}(\theta) & \text{cos}(\theta)
+                \end{bmatrix} \\
+
+        Its ``(A,b,c)`` triple is given by 
+
+        .. math::
+            A &= \begin{bmatrix}
+                    O_2 & U \\
+                    U^{T} & O_2
+                \end{bmatrix} \\
+            b &= O_{4} \\
+            c &= 1
+    """
+
+    def __init__(
+        self,
+        modes: Tuple[int, int],
+        theta: float = 0.0,
+        phi: float = 0.0,
+        theta_trainable: bool = False,
+        phi_trainable: bool = False,
+        theta_bounds: Tuple[Optional[float], Optional[float]] = (None, None),
+        phi_bounds: Tuple[Optional[float], Optional[float]] = (None, None),
+    ):
+        if len(modes) != 2:
+            raise ValueError(f"Expected a pair of modes, found {modes}.")
+
+        super().__init__(modes=modes, name="BSgate")
+        self._add_parameter(make_parameter(theta_trainable, theta, "theta", theta_bounds))
+        self._add_parameter(make_parameter(phi_trainable, phi, "phi", phi_bounds))
+
+    @property
+    def representation(self) -> Bargmann:
```

Comment:
Would this needs a check if the parameters are longer than 1, since we only support BS in two modes